### DISCUSSION TITLE
Beginner - Next JS. 'fs' module not found issue.

### DISCUSSION THREAD

**Discussion by thegitparticle** [vercel/next.js#12124] [Help] [answered]
I am going through the blog tutorial on Next JS site, and encountered this error when trying to implement static rendering. 
```
./lib/posts.js
Module not found: Can't resolve 'fs'
```

**Comment by rafaelalmeidatk** [comment, marked as answer]
Welcome! `fs` is a library exclusive to Node, so you can't use it directly in the browser. But if you import it at the top of the file and use it in a function that only runs in the server, then Next will remove this import from the client bundle, so this code is valid:

```js
import fs from 'fs';

export default function Home(props) {
  return (
    <div className="container">
      ...
    </div>
  );
}

// getStaticProps only runs in the Node side, so it is safe
// to use libraries from Node here
export async function getStaticProps() {
  const fileNames = fs.readdirSync(postsDirectory);

  ...
}
```

In the tutorial these libraries are imported in another file, but a bit later they are utilized inside `getStaticProps` so Next knows they can be removed from the client:

```js
export async function getStaticProps() {
  const allPostsData = getSortedPostsData()
  return {
    props: {
      allPostsData
    }
  }
}
```

**Reply by thegitparticle** [reply]
Thanks. That helped!

**Reply by trav3711** [reply]
This is confusing. I'm facing the same issue. I am importing fs at the top of my posts.js which defines getSortedPropData(). fs is called in that function, and you're telling me I don't need to include it in the file?

index.js
```javascript
import Head from 'next/head'
import Layout, { siteTitle } from '../components/layout'
import utilStyles from '../styles/utils.module.css'
import { getSortedPostsData } from '../lib/posts'

export async function getStaticprops() {
  const allPostsData = getSortedPostsData()
  return {
    props: {
      allPostsData
    }
  }
}

export default function Home({ allPostsData }) {
  return (
    <Layout home>

      <Head>
        <title>{siteTitle}</title>
      </Head>
      <section className={utilStyles.headingMd}>
        <p>Im just a dude who likes to do things</p>
        <p>I like to play chess</p>
      </section>

      <section className={`${utilStyles.headingMd} ${utilStyles.padding1px}`}>
        <h2 className={utilStyles.headingLg}>Blog</h2>
        <ul className={utilStyles.list}>
          {allPostsData.map(({ id, date, title }) => (
            <li className={utilStyles.listItem} key={id}>
              {title}
              <br />
              {id}
              <br />
              {date}
            </li>
          ))}
        </ul>
      </section>

    </Layout>
  )
}
```
posts.js
```javascript
import fs from 'fs'
import path from 'path'
import matter from 'gray-matter'

const postsDirectory = path.join(process.cwd(), 'posts')

export function getSortedPostsData() {
  // Get file names under /posts
  const fileNames = fs.readdirSync(postsDirectory)
  const allPostsData = fileNames.map(fileName => {
    // Remove ".md" from file name to get id
    const id = fileName.replace(/\.md$/, '')

    // Read markdown file as string
    const fullPath = path.join(postsDirectory, fileName)
    const fileContents = fs.readFileSync(fullPath, 'utf8')

    // Use gray-matter to parse the post metadata section
    const matterResult = matter(fileContents)

    // Combine the data with the id
    return {
      id,
      ...matterResult.data
    }
  })
  // Sort posts by date
  return allPostsData.sort((a, b) => {
    if (a.date < b.date) {
      return 1
    } else {
      return -1
    }
  })
}
```

**Reply by LauraCole1900** [reply]
Does that mean that the line in the "Implement getStaticProps" instructions that reads "You should now see the blog data if you access http://localhost:3000." is bs? I literally copied and pasted the code from the tutorial into VSCode and am getting this error.

**Reply by jde** [reply]
Programs are complex. Writing tutorials is hard and there's a lot of room for error. I recommend shifting your mindset from "that is bs" to "what does this error mean." This error is one of the harder ones to reason about since it points to an error somewhere else in your code. I wrote a bit about it elsewhere on this thread: 

https://github.com/vercel/next.js/discussions/12124#discussioncomment-454290

**Reply by netrolite** [reply]
what about the new app directory? How do I use the `fs` module? I'm getting the error `TypeError: fs__WEBPACK_IMPORTED_MODULE_0___default(...).readdirSync is not a function`.  I heard that you can no longer use functions like `getStaticProps`...

**Reply by netrolite** [reply]
I found what the issue was. I simply had a client component trying to render a server component. But if I'm not mistaken, the docs say that your client components should be like the leaves of your server components. In other words, you can nest client components inside server components, but not vice versa.

**Reply by Hostitel** [reply]
> Welcome! `fs` is a library exclusive to Node, so you can't use it directly in the browser. But if you import it at the top of the file and use it in a function that only runs in the server, then Next will remove this import from the client bundle, so this code is valid:
> 
> 
> 
> ```js
> 
> import fs from 'fs';
> 
> 
> 
> export default function Home(props) {
> 
>   return (
> 
>     <div className="container">
> 
>       ...
> 
>     </div>
> 
>   );
> 
> }
> 
> 
> 
> // getStaticProps only runs in the Node side, so it is safe
> 
> // to use libraries from Node here
> 
> export async function getStaticProps() {
> 
>   const fileNames = fs.readdirSync(postsDirectory);
> 
> 
> 
>   ...
> 
> }
> 
> ```
> 
> 
> 
> In the tutorial these libraries are imported in another file, but a bit later they are utilized inside `getStaticProps` so Next knows they can be removed from the client:
> 
> 
> 
> ```js
> 
> export async function getStaticProps() {
> 
>   const allPostsData = getSortedPostsData()
> 
>   return {
> 
>     props: {
> 
>       allPostsData
> 
>     }
> 
>   }
> 
> }
> 
> ```

**Comment by theogravity** [comment]
I'm not sure what I'm doing incorrectly. The sample does work, but when I try to write my own component, I get the error:

```typescript
import { promises } from 'fs'
import React from 'react'
import { join } from 'path'
import firebase from '../utils/firebase'

// Configure FirebaseUI.
const uiConfig = {
  // Redirect to /signedIn after sign in is successful. Alternatively you can provide a callbacks.signInSuccess function.
  signInSuccessUrl: '/signedIn',
  // We will display Google and Facebook as auth providers.
  signInOptions: [
    firebase.auth.GoogleAuthProvider.PROVIDER_ID,
    firebase.auth.EmailAuthProvider.PROVIDER_ID,
  ]
}

export default function LoginOptions ({ css }) {
  React.useEffect(() => {
    // must be required here since this is client-side rendered
    const firebaseui = require('firebaseui')

    // Initialize the FirebaseUI Widget using Firebase.
    const ui = new firebaseui.auth.AuthUI(firebase.auth())

    // The start method will wait until the DOM is loaded.
    ui.start('#firebaseui-auth-container', uiConfig);
  }, [])

  return (
    <>
    <div id="firebaseui-auth-container"></div>
      <style jsx global>{css}</style>
    </>
  )
}

export async function getStaticProps () {
  const css = await promises.readFile(join(process.cwd(), 'public', 'firebaseui.css'))

  return {
    props: {
      css
    }
  }
}
```

**Reply by rafaelalmeidatk** [reply]
You can only use `getStaticProps` from page components (located in the `pages/` folder). FWIW Next.js has [Built-in CSS Support](https://nextjs.org/docs/basic-features/built-in-css-support), so you can import a global `.css` file from `_app.js`

**Reply by theogravity** [reply]
@rafaelalmeidatk oh. thanks

On the second point, I don't want to include that css globally. It's specific to that page only. I'll move the logic to the page level.

**Reply by eddrichjanzzen** [reply]
This answer was what resolved my issue! the component needs to be inside the `pages/` folder for it to work, cheers dude!

**Comment by Morantron** [comment]
i'm getting this error by just adding an `import fs from 'fs'` on the top of page components

```
yarn create next-app
// go to pages/index.js and add import
yarn dev
```

**Reply by rafaelalmeidatk** [reply]
This is the expected behavior, `fs` doesn't exist in the browser. If you want Next.js to remove it from the client bundle you need to use it inside `getServerSideProps`/`getStaticProps`

**Reply by Morantron** [reply]
thanks! this is more or less what was happening in my use case


```javascript
// lib/loadPosts.js

import fs from "fs";

export function getAllPosts() {
  return fs.readdirSync("/path/to/posts");
}

export function getPostBySlug(slug) {
  return fs.readFileSync("/path/to/posts" + slug, "utf8");
}

export function someUnusedFunction() {
  return fs.readdirSync("/tmp");
}

// pages/blog/[slug].js

import { getAllPosts, getPostBySlug, someUnusedFunction } from "lib/loadPosts";

export async function getStaticPaths() {
  const posts = getAllPosts();

  return {
    paths: posts.map(posts => {
      return {
        params: {
          slug: posts.slug
        }
      };
    }),
    fallback: false
  };
}

export async function getStaticProps({ params }) {
  const post = getPostBySlug(params.slug);

  return {
    props: { post }
  };
}

const BlogPost = ({ post }) => {
  return <pre>{JSON.stringify(post, null, 2)}</pre>;
};
```

I was using a logic similar to the `blog-starter` one in the examples folder, but imported a function that I wasn't really using.

A more explicit error message in this case would be great for DX, but maybe it's hard to detect.

**Reply by beshur** [reply]
I have the same code, but it also does not work

**Comment by patarapolw** [comment]
Now, I create a TS file just for typing. It's actually this, and it throws error. It does not ignore `fs` (and `path`), and just tree-shaking...

```ts
import fs from 'fs'
import yaml from 'js-yaml'
import path from 'path'
import * as z from 'zod'

export const zRemark42 = () => z.object({
  host: z.string(),
  siteId: z.string(),
  locale: z.string().optional()
})

export type IRemark42 = z.infer<ReturnType<typeof zRemark42>>

export const zAuthor = () => z.object({
  name: z.string(),
  email: z.string().optional(),
  url: z.string().optional()
})

export type IAuthor = z.infer<ReturnType<typeof zAuthor>>

export const zTabs = () => z.array(z.object({
  name: z.string(),
  to: z.string().optional(),
  href: z.string().optional()
}))

export type ITabs = z.infer<ReturnType<typeof zTabs>>

export const zSidebar = () => z.object({
  twitter: z.string().optional(),
  tagCloud: z.object({
    excluded: z.array(z.string()).optional()
  }).optional()
})

export type ISidebar = z.infer<ReturnType<typeof zSidebar>>

export const zSocial = () => z.object({
  twitter: z.string().optional(),
  reddit: z.string().optional(),
  quora: z.string().optional(),
  github: z.string().optional()
})

export type ISocial = z.infer<ReturnType<typeof zSocial>>

export const zTheme = () => z.object({
  title: z.string(),
  banner: z.string().optional(),
  baseUrl: z.string().optional(),
  description: z.string().optional(),
  keywords: z.array(z.string()).optional(),
  tabs: zTabs().optional(),
  author: zAuthor(),
  social: zSocial().optional(),
  sidebar: zSidebar().optional(),
  analytics: z.object({
    plausible: z.string().optional()
  }).optional(),
  comments: z.object({
    remark42: zRemark42().optional()
  }).optional()
})

export const getTheme = () => zTheme().parse(yaml.safeLoad(fs.readFileSync(path.join(__dirname, '../theme.yml'), 'utf8')))

export type ITheme = z.infer<ReturnType<typeof zTheme>>
```

**Comment by danielzzz** [comment]
it seems that I cannot use fs event in getStaticProps, the compiler renders an error just when importing fs, even if it's not used at all..

**Reply by jde** [reply]
I had a similar issue. It turns out that I had a conceptually unrelated error in which I was requiring an unexported function from a module. When I resolved that issue, the cannot require 'fs' error went away. I can't account for why these two are connected, but double check the rest of your code.

**Reply by jde** [reply]
After a few weeks with next I developed an understanding of what's happening here. This is a classic unintuitive consequence of having code that runs on the server and in the browser in the same file / code base. `fs` is a node file system library. It exists on the server but not in the browser (since there's not filesystem in the browser.)

If there are errors in your code, the compiler will not be able to understand what to run in the browser and on the server. When this happens, you may see the `fs` error before your browser gets to the error that caused it. This can be very confusing, but once you know that the error is not being reported correctly, it's just a matter of taking a close look at what you did last to make sure you don't have other errors. Good IDE plugins can help with this as well.

**Reply by Coder-Jedi** [reply]
Thanks a lot! I was facing the same issue, actually it was a totally unrelated error to that being  shown in the console.

**Reply by tracyhenry** [reply]
For me, I accidentally typed `getPostData` as `getPostdata` when importing, which caused next.js to include `lib/posts.js` in the client bundle.

**Reply by johnbarhorst** [reply]
This information fixed it all for me. I was using type script and hadn't typed my stuff yet. Thanks!

**Reply by Cerity06** [reply]
Thank you ! At last the information that has resolved my issue :)

**Reply by DanB-Web** [reply]
Thanks @jde, that was helpful

**Comment by richellyitalo** [comment]
I made this way:

```js
import path from 'path';

function HomePage(props) {
  const { products, filePath } = props;

  return (
    <ul>
      {products.map((product) => (
        <li key={product.id}>{product.title}</li>
      ))}
    </ul>
  );
}

export async function getStaticProps(context) {
  const fs = require('fs/promises'); // LOOK HERE

  const filePath = path.join(process.cwd(), 'data', 'dummy-data.json');
  const fileData = await fs.readFile(filePath);
  const data = JSON.parse(fileData);

  return {
    props: {
      products: data.products,
    },
  };
}

export default HomePage;

```

**Reply by yddc902** [reply]
you made my day buddy, thanks

**Reply by orcololo** [reply]
What about typeScript version? 
This is killing me. Hahaha

**Comment by tatimblin** [comment]
I had the same issue, apparently the issue is that `fs` is a Node module and can only run server side. Next.js accounts for this letting it run in `getStaticProps()`, but the error still surfaces you can resolve the error by creating a `nuxt.config.js` and setting `fs` to be an empty module.

```

module.exports = {
  webpack: (config, { isServer }) => {
    // Fixes npm packages that depend on `fs` module
    if (!isServer) {
      config.node = {
        fs: 'empty'
      }
    }

    return config
  }
}
```

I learned this from [this article](https://javascript.plainenglish.io/understand-code-splitting-in-next-js-by-solving-a-common-error-db61a9b1b1a1), where someone had the same issue

**Reply by Ajay-056** [reply]
> I had the same issue, apparently the issue is that `fs` is a Node module and can only run server side. Next.js accounts for this letting it run in `getStaticProps()`, but the error still surfaces you can resolve the error by creating a `nuxt.config.js` and setting `fs` to be an empty module.
> 
> ```
> 
> module.exports = {
>   webpack: (config, { isServer }) => {
>     // Fixes npm packages that depend on `fs` module
>     if (!isServer) {
>       config.node = {
>         fs: 'empty'
>       }
>     }
> 
>     return config
>   }
> }
> ```
> 
> I learned this from [this article](https://javascript.plainenglish.io/understand-code-splitting-in-next-js-by-solving-a-common-error-db61a9b1b1a1), where someone had the same issue

This is not working for me anyhow the 'fs' module can't be resolved error is gone and now getting another error tried quite a solutions 

ValidationError: Invalid configuration object. Webpack has been initialized using a configuration object that does not match the API schema.
 - configuration[0].node has an unknown property 'fs'. These properties are valid:
   object { __dirname?, __filename?, global? }
   -> Options object for node compatibility features.
    at validate (C:\Users\*\Desktop\whatsapp-clone-nextjs\node_modules\next\dist\compiled\schema-utils3\index.js:1:153657)
    at validateSchema (C:\Users\*\Desktop\whatsapp-clone-nextjs\node_modules\next\dist\compiled\webpack\bundle5.js:137945:2)
    at create (C:\Users\*\Desktop\whatsapp-clone-nextjs\node_modules\next\dist\compiled\webpack\bundle5.js:141384:24)
    at webpack (C:\Users\*\Desktop\whatsapp-clone-nextjs\node_modules\next\dist\compiled\webpack\bundle5.js:141426:32)
    at f (C:\Users\*\Desktop\whatsapp-clone-nextjs\node_modules\next\dist\compiled\webpack\bundle5.js:98978:16)
    at HotReloader.start (C:\Users\*\Desktop\whatsapp-clone-nextjs\node_modules\next\dist\server\hot-reloader.js:18:415)
    at async DevServer.prepare (C:\Users\*\Desktop\whatsapp-clone-nextjs\node_modules\next\dist\server\next-dev-server.js:16:453)       
    at async C:\Users\*\Desktop\whatsapp-clone-nextjs\node_modules\next\dist\cli\next-dev.js:22:1 {
  errors: [
    {
      keyword: 'anyOf',
      dataPath: '[0].node',
      schemaPath: '#/anyOf',
      params: {},
      message: 'should match some schema in anyOf',
      schema: [Array],
      parentSchema: [Object],
      data: [Object],
      children: [Array]
    }
  ],
  schema: {
    definitions: {
      Amd: [Object],
      AssetFilterItemTypes: [Object],
      AssetFilterTypes: [Object],
      AssetGeneratorDataUrl: [Object],
      AssetGeneratorDataUrlFunction: [Object],
      AssetGeneratorDataUrlOptions: [Object],
      AssetGeneratorOptions: [Object],
      AssetInlineGeneratorOptions: [Object],
      AssetModuleFilename: [Object],
      AssetParserDataUrlFunction: [Object],
      AssetParserDataUrlOptions: [Object],
      AssetParserOptions: [Object],
      AssetResourceGeneratorOptions: [Object],
      AuxiliaryComment: [Object],
      Bail: [Object],
      CacheOptions: [Object],
      CacheOptionsNormalized: [Object],
      Charset: [Object],
      ChunkFilename: [Object],
      ChunkFormat: [Object],
      ChunkLoadTimeout: [Object],
      ChunkLoading: [Object],
      ChunkLoadingGlobal: [Object],
      ChunkLoadingType: [Object],
      Clean: [Object],
      CleanOptions: [Object],
      CompareBeforeEmit: [Object],
      Context: [Object],
      CrossOriginLoading: [Object],
      Dependencies: [Object],
      DevServer: [Object],
      DevTool: [Object],
      DevtoolFallbackModuleFilenameTemplate: [Object],
      DevtoolModuleFilenameTemplate: [Object],
      DevtoolNamespace: [Object],
      EmptyGeneratorOptions: [Object],
      EmptyParserOptions: [Object],
      EnabledChunkLoadingTypes: [Object],
      EnabledLibraryTypes: [Object],
      EnabledWasmLoadingTypes: [Object],
      Entry: [Object],
      EntryDescription: [Object],
      EntryDescriptionNormalized: [Object],
      EntryDynamic: [Object],
      EntryDynamicNormalized: [Object],
      EntryFilename: [Object],
      EntryItem: [Object],
      EntryNormalized: [Object],
      EntryObject: [Object],
      EntryRuntime: [Object],
      EntryStatic: [Object],
      EntryStaticNormalized: [Object],
      EntryUnnamed: [Object],
      Environment: [Object],
      Experiments: [Object],
      ExternalItem: [Object],
      ExternalItemFunctionData: [Object],
      ExternalItemValue: [Object],
      Externals: [Object],
      ExternalsPresets: [Object],
      ExternalsType: [Object],
      FileCacheOptions: [Object],
      Filename: [Object],
      FilenameTemplate: [Object],
      FilterItemTypes: [Object],
      FilterTypes: [Object],
      GeneratorOptionsByModuleType: [Object],
      GlobalObject: [Object],
      HashDigest: [Object],
      HashDigestLength: [Object],
      HashFunction: [Object],
      HashSalt: [Object],
      HotUpdateChunkFilename: [Object],
      HotUpdateGlobal: [Object],
      HotUpdateMainFilename: [Object],
      IgnoreWarnings: [Object],
      IgnoreWarningsNormalized: [Object],
      Iife: [Object],
      ImportFunctionName: [Object],
      ImportMetaName: [Object],
      InfrastructureLogging: [Object],
      JavascriptParserOptions: [Object],
      Layer: [Object],
      Library: [Object],
      LibraryCustomUmdCommentObject: [Object],
      LibraryCustomUmdObject: [Object],
      LibraryExport: [Object],
      LibraryName: [Object],
      LibraryOptions: [Object],
      LibraryType: [Object],
      Loader: [Object],
      MemoryCacheOptions: [Object],
      Mode: [Object],
      ModuleFilterItemTypes: [Object],
      ModuleFilterTypes: [Object],
      ModuleOptions: [Object],
      ModuleOptionsNormalized: [Object],
      Name: [Object],
      NoParse: [Object],
      Node: [Object],
      NodeOptions: [Object],
      Optimization: [Object],
      OptimizationRuntimeChunk: [Object],
      OptimizationRuntimeChunkNormalized: [Object],
      OptimizationSplitChunksCacheGroup: [Object],
      OptimizationSplitChunksGetCacheGroups: [Object],
      OptimizationSplitChunksOptions: [Object],
      OptimizationSplitChunksSizes: [Object],
      Output: [Object],
      OutputModule: [Object],
      OutputNormalized: [Object],
      Parallelism: [Object],
      ParserOptionsByModuleType: [Object],
      Path: [Object],
      Pathinfo: [Object],
      Performance: [Object],
      PerformanceOptions: [Object],
      Plugins: [Object],
      Profile: [Object],
      PublicPath: [Object],
      RawPublicPath: [Object],
      RecordsInputPath: [Object],
      RecordsOutputPath: [Object],
      RecordsPath: [Object],
      Resolve: [Object],
      ResolveAlias: [Object],
      ResolveLoader: [Object],
      ResolveOptions: [Object],
      ResolvePluginInstance: [Object],
      RuleSetCondition: [Object],
      RuleSetConditionAbsolute: [Object],
      RuleSetConditionOrConditions: [Object],
      RuleSetConditionOrConditionsAbsolute: [Object],
      RuleSetConditions: [Object],
      RuleSetConditionsAbsolute: [Object],
      RuleSetLoader: [Object],
      RuleSetLoaderOptions: [Object],
      RuleSetLogicalConditions: [Object],
      RuleSetLogicalConditionsAbsolute: [Object],
      RuleSetRule: [Object],
      RuleSetRules: [Object],
      RuleSetUse: [Object],
      RuleSetUseItem: [Object],
      ScriptType: [Object],
      SnapshotOptions: [Object],
      SourceMapFilename: [Object],
      SourcePrefix: [Object],
      StatsOptions: [Object],
      StatsValue: [Object],
      StrictModuleErrorHandling: [Object],
      StrictModuleExceptionHandling: [Object],
      Target: [Object],
      TrustedTypes: [Object],
      UmdNamedDefine: [Object],
      UniqueName: [Object],
      WarningFilterItemTypes: [Object],
      WarningFilterTypes: [Object],
      WasmLoading: [Object],
      WasmLoadingType: [Object],
      Watch: [Object],
      WatchOptions: [Object],
      WebassemblyModuleFilename: [Object],
      WebpackOptionsNormalized: [Object],
      WebpackPluginFunction: [Object],
      WebpackPluginInstance: [Object]
    },
    title: 'WebpackOptions',
    description: 'Options object as provided by the user.',
    type: 'object',
    additionalProperties: false,
    properties: {
      amd: [Object],
      bail: [Object],
      cache: [Object],
      context: [Object],
      dependencies: [Object],
      devServer: [Object],
      devtool: [Object],
      entry: [Object],
      experiments: [Object],
      externals: [Object],
      externalsPresets: [Object],
      externalsType: [Object],
      ignoreWarnings: [Object],
      infrastructureLogging: [Object],
      loader: [Object],
      mode: [Object],
      module: [Object],
      name: [Object],
      node: [Object],
      optimization: [Object],
      output: [Object],
      parallelism: [Object],
      performance: [Object],
      plugins: [Object],
      profile: [Object],
      recordsInputPath: [Object],
      recordsOutputPath: [Object],
      recordsPath: [Object],
      resolve: [Object],
      resolveLoader: [Object],
      snapshot: [Object],
      stats: [Object],
      target: [Object],
      watch: [Object],
      watchOptions: [Object]
    }
  },
  headerName: 'Webpack',
  baseDataPath: 'configuration',
  postFormatter: [Function: postFormatter]
}
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! whatsapp-clone-nextjs@0.1.0 dev: `next dev`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the whatsapp-clone-nextjs@0.1.0 dev script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\*\AppData\Roaming\npm-cache\_logs\2021-07-19T03_58_05_317Z-debug.log

**Reply by bonesoul** [reply]
same, any solutions?

**Reply by matthill8286** [reply]
@bonesoul 

```
  webpack5: true,
  webpack            : config => {
    config.resolve.fallback = {...config.resolve.fallback, fs: false}

    return config
  },
  ```

**Comment by progresssh** [comment]
To those who might not find a solution in the answers above

Verify your `_app.js`

It needs to be as written below
```js
import "../styles/global.css";

export default function App({ Component, pageProps }) {
  return <Component {...pageProps} />;
}
```

**NOT** like this
```js
import "../styles/global.css";

export default function App({ Component }, { pageProps }) {
  return <Component {...pageProps} />;
}
```
which was my problem, in this example `pageProps` doesn't get any `props` to pass to `page`, thus `pageProps` is an empty object.

**Comment by sapeol** [comment]
i had the same issue and i fixed it by `npm remove fs` because fs is also a [security package](https://www.npmjs.com/package/fs)

**Comment by ksajjan** [comment]
After figuring out for 2 hours finally It resolved. I faced the issue because of  wrong function name. Instead of `getStaticPaths`, I was putting a different name `getStatticPaths`.

```
export async function getStaticPaths() {
    const  paths = getAllPostIds();
    return {
        paths,
        fallback: false
    };
}
```
**Check properly the name of functions.**

**Comment by front-end-apps** [comment]
Hi Team, 

How can I writeJsonFile in next.js / react ?

**Reply by cwtuan** [reply]
@front-end-apps Refer to solution https://github.com/vercel/next.js/discussions/12124#discussioncomment-4012110

**Comment by cwtuan** [comment]
Don't use `fs` in the `pages` directory, since next.js suppose that files in `pages` directory are running in browser environment.

You could put the util file which uses `fs` to other directory such as `/core`

Then `require` the util in `getStaticProps` which runs in node.js environment.

```tsx
// /pages/myPage/index.tsx
import View from './view';
export default View;

export async function getStaticProps() {
  const util = require('core/some-util-uses-fs').default; // getStaticProps runs in nodes
  const data = await util.getDataFromDisk();
  return {
    props: {
      data,
    },
  };
}

```

**Comment by adityanithariya** [comment]
I'm using winston for logging at server side, and toast utility at client side, tried a dynamic import of winston, and file where logger is implemented, but its still giving fs error, tried `next.config.js` webpack fn change fix, but its not working, using Next.js 13 App Router.
@ijjk Please help!
Thanks in Advance!!
